### PR TITLE
[`Styling`] stylify using ruff

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Removes our dependency on `black` in favor of rust:
- faster `make style` I'll try to get numbers but that one is pretty obvious 
- not need to use doc-builder's black styling as the doc is styled in ruff 
- faster overall CI to check the repo quality

### CODE REVIEW

File: examples/research_projects/movement-pruning/Saving_PruneBERT.ipynb [hunk 1]
```diff
@@ -533,9 +533,9 @@
     "\n",
     "for name, param in reconstructed_qtz_st.items():\n",
     "    assert type(param) == type(qtz_st[name]), name\n",
-    "    if type(param) == torch.Tensor:\n",
+    "    if isinstance(param, torch)rch).Tensor:\n",
```
Comment (ydshieh) [inline examples/research_projects/movement-pruning/Saving_PruneBERT.ipynb:536]: something strange here ..?

File: pyproject.toml [hunk 2]
```diff
@@ -18,6 +14,19 @@ line-length = 119
...
+quote-style = "double"
+
+# Like Black, indent with spaces, rather than tabs.
+indent-style = "space"
+
+# Like Black, respect magic trailing commas.
+skip-magic-trailing-comma = false
+
+# Like Black, automatically detect the appropriate line ending.
+line-ending = "auto"
```
Comment (LysandreJik) [inline pyproject.toml:28]: cool, thanks for keeping this explicit

**Review by LysandreJik — APPROVED** [review]
Very cool! The changes look good to me.

There's one thing that we should do here I think: we should look for the fmt: on/off and remove them/update them to their newer counterparts.

We usually look for them when searching how to disable the format, so having examples of how to do so would be nice. Ensuring that we have no "old" formatting statements would be cleaner as well.

File: .circleci/create_circleci_config.py [hunk 3]
```diff
@@ -517,7 +516,7 @@ def job_name(self):
         "pip install --upgrade --upgrade-strategy eager pytest pytest-sugar",
         "pip install -U --upgrade-strategy eager natten",
         "find -name __pycache__ -delete",
-        "find . -name \*.pyc -delete",
+        "find . -name \\*.pyc -delete",
```
Comment (ydshieh) [inline .circleci/create_circleci_config.py:519]: I would personally check if this new (format) command would work well on a machine's bash commandline.

File: docs/source/en/tasks/semantic_segmentation.md [hunk 4]
```diff
@@ -245,7 +245,7 @@ logits first, and then reshaped to match the size of the labels before you can c
 ...             reduce_labels=False,
 ...         )
 ...         for key, value in metrics.items():
-...             if type(value) is np.ndarray:
+...             if isinstance(value, np.ndarray)
```
Comment (ydshieh) [inline docs/source/en/tasks/semantic_segmentation.md:248]: I feel a bit strange here `:` is removed. Is this expected?

File: tests/pipelines/test_pipelines_token_classification.py [hunk 5]
```diff
@@ -561,7 +561,18 @@ def test_aggregation_strategy(self):
...
+                        0,
+                        0,
+                        0,
+                        0,
+                        0,
+                        0.9986497163772583,
+                        0,
+                        0,
+                    ]
+                ),
```
Comment (ydshieh) [inline tests/pipelines/test_pipelines_token_classification.py:575]: Guess this is not what we love to have, and related to Lysandre's comment I believe

File: setup.py [hunk 6]
```diff
@@ -156,7 +155,7 @@
     "rhoknp>=1.1.0,<1.3.1",
     "rjieba",
     "rouge-score!=0.0.7,!=0.0.8,!=0.1,!=0.1.1",
-    "ruff>=0.0.241,<=0.0.259",
+    "ruff>=0.1.5,<=0.2",
```
Comment (ydshieh) [inline setup.py:158]: Could you elaborate a bit on this choice?

File: utils/check_copies.py [hunk 7]
```diff
@@ -238,11 +247,10 @@ def blackify(code: str) -> str:
     """
     has_indent = len(get_indent(code)) > 0
     if has_indent:
+        pass
```
Comment (ydshieh) [inline utils/check_copies.py:250]: pass will still run the nex line , right?

File: utils/check_copies.py [hunk 8]
```diff
@@ -226,9 +225,19 @@ def get_indent(code: str) -> str:
...
 
 
-def blackify(code: str) -> str:
+async def run_ruff(code):
+    command = ["ruff", "format", "-", "--config", "pyproject.toml", "--silent"]
+    process = await asyncio.create_subprocess_exec(
+        *command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE
+    )
+    stdout, _ = await process.communicate(input=code.encode())
+    return stdout.decode()
```
Comment (ydshieh) [inline utils/check_copies.py:234]: would love to hear the motivation to make is async 🙏

**Review by ydshieh — APPROVED** [review]
Thanks for this awesome work! Having a few question, but sure LGTM!

File: docs/source/en/tasks/semantic_segmentation.md (hunk 4 above)
Comment (ArthurZucker) [reply docs/source/en/tasks/semantic_segmentation.md:248]: nice catch! Thanks

File: utils/check_copies.py (hunk 7 above)
Comment (ArthurZucker) [reply utils/check_copies.py:250]: Yes, but I'll just remove the pass it's useless

File: utils/check_copies.py (hunk 8 above)
Comment (ArthurZucker) [reply utils/check_copies.py:234]: Followed what was done in huggingface hub, nor real reasons apart trying to make it faster. Though it won't be run in parallel. Did not benchmark both

File: setup.py (hunk 6 above)
Comment (ArthurZucker) [reply setup.py:158]: Latest version for better performance and unified api I'd say. While <=2 to have enough rolling in case they change things

Comment (ydshieh) [reply setup.py:158]: Oh, I misread  the numbers. Got it!

File: utils/check_copies.py (hunk 8 above)
Comment (ydshieh) [reply utils/check_copies.py:234]: Well, as here we don't really use the concurrency (which is the advantage of using `coroutine`), it looks somehow overwhelming to use `async` stuff here.

But my real concern is: here we spawn a new process whenever there is a `# Copied from` statement, which is an expensive op. Previously, `blackify` uses `black.format_str` which is inside the same python process, despite black itself is slow.

I checked the run (check_copies) with this PR and on `main`: 44 sec. on this PR and 32 sec. on `main` .

See [run with this PR](https://app.circleci.com/pipelines/github/huggingface/transformers/78517/workflows/e0cbb6e0-2f1a-4349-a8ea-873d81958a72/jobs/1001612) and [run with main](https://app.circleci.com/pipelines/github/huggingface/transformers/78474/workflows/b3b421a1-aefc-48c5-bcb4-53bff8bc3560/jobs/1000961).

I am afraid, at least for `check_copies.py`, we lose the advantage of using `ruff format` due to the many new processes being created each time.

Comment (ArthurZucker) [reply utils/check_copies.py:234]: Well I can do not async, no worries. Black is just too complicated, and with ruff you don't need to use doc-builder on top of it. 
Now differences can also come from modified files as well but let's remove black because if you take the `make style` time into account it's already worth it
